helix-view/src/register.rs

```diff
@@ -0,0 +1,235 @@
+///
+/// * Black hole (`_`): all values read and written are discarded
+/// * Selection indices (`#`): index number of each selection starting at 1
+/// * Selection contents (`.`)
+/// * Document path (`%`): filename of the current buffer
+/// * System clipboard (`*`)
+/// * Primary clipboard (`+`)
+#[derive(Debug)]
+pub struct Registers {
+    inner: HashMap<char, Vec<String>>,
+    clipboard_provider: Box<dyn ClipboardProvider>,
+}
+
+impl Default for Registers {
+    fn default() -> Self {
+        Self {
+            inner: Default::default(),
+            clipboard_provider: get_clipboard_provider(),
+        }
+    }
+}
+
+// Some special registers must allocate their values while others and regular
+// registers can hand out borrowed values.
+type RegisterValues<'a> = Box<dyn ExactSizeIterator<Item = Cow<'a, str>> + 'a>;
+
+impl Registers {
+    pub fn read<'a>(&'a self, name: char, editor: &'a Editor) -> Option<RegisterValues<'a>> {
+        match name {
+            '_' => Some(Box::new(iter::empty())),
+            '#' => {
+                let (view, doc) = current_ref!(editor);
+                let selections = doc.selection(view.id).len();
+                // ExactSizeIterator is implemented for Range<usize> but
+                // not RangeInclusive<usize>.
+                Some(Box::new(
+                    (0..selections).map(|i| i.saturating_add(1).to_string().into()),
+                ))
+            }
+            '.' => {
+                let (view, doc) = current_ref!(editor);
+                let text = doc.text().slice(..);
+                Some(Box::new(doc.selection(view.id).fragments(text)))
+            }
+            '%' => {
+                let doc = doc!(editor);
+
+                let path = doc
+                    .path()
+                    .as_ref()
+                    .map(|p| p.to_string_lossy())
+                    .unwrap_or_else(|| SCRATCH_BUFFER_NAME.into());
+
+                Some(Box::new(iter::once(path)))
+            }
+            '*' | '+' => Some(read_from_clipboard(
+                self.clipboard_provider.as_ref(),
+                self.inner.get(&name),
+                match name {
+                    '*' => ClipboardType::Clipboard,
+                    '+' => ClipboardType::Selection,
+                    _ => unreachable!(),
+                },
+            )),
+            _ => self
+                .inner
+                .get(&name)
+                .map(|values| Box::new(values.iter().map(Cow::from).rev()) as RegisterValues),
+        }
+    }
+
+    pub fn write(&mut self, name: char, mut values: Vec<String>) -> Result<()> {
+        match name {
+            '_' => Ok(()),
+            '#' | '.' | '%' => Err(anyhow::anyhow!("Register {name} does not support writing")),
+            '*' | '+' => {
+                self.inner.insert(name, values);
+                self.clipboard_provider.set_contents(
+                    self.inner[&name].join(NATIVE_LINE_ENDING.as_str()),
```

Comment:
this also needs to be reversesd before being inserted (but not before being set to the
```suggestion
                let clipboard = values.join(NATIVE_LINE_ENDING.as_str());
                values.reverse();
                self.inner.insert(name, values);
                self.clipboard_provider.set_contents(
                    clipboard,
```